devtools/src/main/scala/kalix/devtools/impl/DockerComposeUtils.scala

```diff
@@ -35,22 +37,62 @@ object DockerComposeUtils {
 
       throw new IllegalArgumentException(s"File '$file' does not exist. $extraMsg")
     }
+  }
 
-  def start(file: String, userFunctionPort: Int): Unit = {
-
-    execIfFileExists(file) {
-      Process(s"docker-compose -f $file up", None, "USER_FUNCTION_PORT" -> userFunctionPort.toString).run
+  // read the file once and cache the lines
+  // we will need to iterate over it more than once
+  private lazy val lines: Seq[String] = {
+    if (Files.exists(Paths.get(file))) {
+      val src = Source.fromFile(file)
+      try {
+        src.getLines.toList
+      } finally {
+        src.close()
+      }
+    } else {
+      Seq.empty
     }
+  }
+
+  def start(): Unit = {
+    execIfFileExists {
+      Process(s"docker-compose -f $file up", None).run
 
-    // shutdown hook to down containers when jvm exits
-    sys.addShutdownHook {
-      execIfFileExists(file) {
-        stop(file)
+      // shutdown hook to down containers when jvm exits
+      sys.addShutdownHook {
+        execIfFileExists {
+          stop()
+        }
       }
     }
   }
 
-  private def stop(file: String): Unit = {
+  private def stop(): Unit = {
     s"docker compose -f $file stop".!
   }
+
+  def readUserFunctionPort: Int = {
+    envVar
+      .get("USER_FUNCTION_PORT")
+      .map(_.toInt)
+      .orElse(readUserFunctionPortFromFile)
+      .getOrElse(8080)
+  }
+
+  private val ExtractPort = """\D+:*(\d+)""".r
```

Comment:
Is this meant to not accept something like `127.0.0.1:8899` ? I thought that would need to be accepted..
Also, what is the purpose of that `*`?

I'm wondering if something like `.+:(\d+)` does not suffice..

Reply:
> Also, what is the purpose of that `*`?

Oh, I just saw now.. because its usually defined like `${USER_FUNCTION_PORT:-8081}`... I was looking at that for the normal types of input local:port... sorry. Let me see again with fresh eyes about the `}`. :)

Reply:
yep, the intention is to extract the port from lines defined as:

* `USER_FUNCTION_PORT:${USER_FUNCTION_PORT:-8080}`
* `USER_FUNCTION_PORT:8080`
* `USER_FUNCTION_PORT:${MY_OWN_VARIABLE_NAME:-8080}`

Reply:
Arguably, we could fix it to match `USER_FUNCTION_PORT:` at start of line.

And, also the last option is not useful because we won't be able to read `MY_OWN_VARIABLE_NAME` anyway. 

This brings us down to the following options:

* `USER_FUNCTION_PORT:${USER_FUNCTION_PORT:-8080}`
* `USER_FUNCTION_PORT:8080`

Reply:
I will add an extra test case to match `USER_FUNCTION_PORT:8080`

Reply:
ok, to remove the hack you can use `\D+:*(\d+)\}?` or `\D+:*(\d+).?`